rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -282,15 +322,183 @@ object ExecutionBatch extends Loggable {
   )
 
   /*
-   * Utility method to factor out common logging task and be assured that
-   * the log message is actually sync with the info type.
+   * How long time a run is valid AFTER AN UPDATE (i.e, not in permanent regime).
+   * This is shorter than runValidityTime, because a config update IS a change and
+   * force to send reports in all case.
+   */
+  def updateValidityDuration(runIntervalInfo: ResolvedAgentRunInterval) = runIntervalInfo.interval.plus(GRACE_TIME_PENDING)
+
+  /*
+   * How long time a run is valid before receiving any report (but not after an update)
    */
-  private[this] def runType(traceMessage: String, runType: RunAndConfigInfo)(implicit nodeId: NodeId): RunAndConfigInfo = {
-    val msg = if(traceMessage.trim.size == 0) "" else ": " + traceMessage
-    ComplianceDebugLogger.node(nodeId).trace(s"Run config for node ${nodeId.value}: ${runType.logName} ${msg}")
-    runType
+  def runValidityDuration(runIntervalInfo: ResolvedAgentRunInterval, complianceMode: ComplianceMode) = complianceMode.mode match {
+    case ChangesOnly =>
+      //expires after run*heartbeat period - we need an other run before that.
+      val heartbeat = Duration.standardMinutes((runIntervalInfo.interval.getStandardMinutes * runIntervalInfo.heartbeatPeriod ))
+      heartbeat.plus(GRACE_TIME_PENDING)
+    case FullCompliance | ReportsDisabled =>
+      updateValidityDuration(runIntervalInfo)
   }
 
+  /*
+   * A data structure with a valide state and its validity interval
+   * (actually the uper bound only, the start of the interval is the run exec time)
+   */
+  final case class RunAndConfigInfoInterval(
+      validState  : RunAndConfigInfo with RunKind.Valid
+    , expiration  : DateTime
+    , invalidState: RunAndConfigInfo with RunKind.Invalid
+  )
+
+  /*
+   * Method that, given a "run and config info" object, calculates what is the run state
+   * given node config info.
+   *
+   * The method returns either an error (no expected reports, bad config id, no reports at all...)
+   * or a datastructure with the Correct state and its validity interval, and the next state after the
+   * interval (before interval does not make sense, the begining of the interval is the run execution time)
+   */
+  def computeNodeRunAndConfigInfo(
+      runInfos         : AgentRunWithNodeConfig
+    , currentNodeConfig: Option[NodeExpectedReports]
```

Comment:
here, the currentNodeConfig is not the current one, but the one we are computed
so maybe use simply nodeConfig (it had to think for 5 minutes to come up with such a lousy name)

Reply:
Hummm, I just move that part around, but you are right, the meaning of "current" (in the sense of "the last one available") is not at all the same now. I will look to it.